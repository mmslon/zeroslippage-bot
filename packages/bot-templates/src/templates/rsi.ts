import { z } from "zod";
import { buy, cancelSmartTrade, IBotConfiguration, sell, TBotContext, useRSI } from "@opentrader/bot-processor";
import { logger } from "@opentrader/logger";

/**
 * Inspired by https://github.com/askmike/gekko/blob/develop/strategies/RSI.js
 * @param ctx - Bot context
 */
export function* rsi(ctx: TBotContext<RsiParams, RsiState>) {
  const {
    config: { settings: params },
    state,
  } = ctx;

  if (ctx.onStart) {
    logger.info(params, "[RSI] Bot started with params");
    return;
  }

  if (ctx.onStop) {
    logger.info("[RSI] Bot stopped");
    yield cancelSmartTrade();
    return;
  }

  if (!state.trend) {
    state.trend = {
      direction: "none",
      duration: 0,
      persisted: false,
      adviced: false,
    };

    logger.info(state, "[RSI] State initialized");
  }

  const rsi: number = yield useRSI(params.periods);
  logger.info(`[RSI] RSI value: ${rsi}`);

  if (rsi > params.high) {
    // new trend detected
    if (state.trend.direction !== "high") {
      state.trend = {
        duration: 0,
        persisted: false,
        direction: "high",
        adviced: false,
      };
    }

    state.trend.duration++;

    logger.info(`[RSI] In high since ${state.trend.duration} candle(s)`);

    if (state.trend.duration >= params.persistence) {
      state.trend.persisted = true;
    }

    if (state.trend.persisted && !state.trend.adviced) {
      state.trend.adviced = true;

      logger.info("[RSI] Advised to SELL");
      yield sell({
        quantity: params.quantity,
        orderType: "Market",
      });
    }
  } else if (rsi < params.low) {
    // new trend detected
    if (state.trend.direction !== "low") {
      state.trend = {
        duration: 0,
        persisted: false,
        direction: "low",
        adviced: false,
      };
    }

    state.trend.duration++;

    logger.info(`[RSI] In low since ${state.trend.duration} candle(s)`);

    if (state.trend.duration >= params.persistence) {
      state.trend.persisted = true;
    }

    if (state.trend.persisted && !state.trend.adviced) {
      state.trend.adviced = true;

      logger.info("[RSI] Advised to BUY");
      yield buy({
        quantity: params.quantity,
        orderType: "Market",
      });
    }
  } else {
    logger.info("[RSI] In no trend");
  }
}

rsi.displayName = "RSI Strategy";
rsi.schema = z.object({
  high: z.number().min(0).max(100).default(70).describe("Sell when RSI is above this value"),
  low: z.number().min(0).max(100).default(30).describe("Buy when RSI is below this value"),
  periods: z.number().positive().default(14).describe("RSI period"),
  persistence: z.number().positive().default(1).describe("Number of candles to persist in trend before buying/selling"),
  quantity: z.number().positive().default(0.0001).describe("Quantity to buy/sell"),
});

rsi.requiredHistory = 15;
rsi.timeframe = ({ timeframe }: IBotConfiguration) => timeframe;
rsi.runPolicy = {
  onCandleClosed: true,
};
rsi.watchers = {
  watchCandles: ({ symbol }: IBotConfiguration) => symbol,
};

type RsiState = {
  trend?: {
    direction: "high" | "low" | "none";
    duration: number;
    persisted: boolean;
    adviced: boolean;
  };
};

export type RsiParams = IBotConfiguration<z.infer<typeof rsi.schema>>;
