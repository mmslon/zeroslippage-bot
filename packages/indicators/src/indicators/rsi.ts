/**
 * Copyright 2024 bludnic
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Repository URL: https://github.com/bludnic/opentrader
 */
import type { ICandlestick } from "@opentrader/types";
import { RSI } from "technicalindicators";
import { IndicatorError } from "../utils/indicator.error.js";

type RsiParams = {
  periods: number;
};

/**
 * Calculate the Relative Strength Index (RSI) for a given set of candles.
 *
 * @param params - RSI parameters
 * @param candles - The candles to calculate the RSI for
 * @returns The RSI values. If there are not enough data points to calculate the RSI, the initial values will be NaN.
 */
export async function rsi(
  params: RsiParams,
  candles: ICandlestick[],
): Promise<number[]> {
  const prices = candles.map((candle) => candle.close);

  if (params.periods < 2) {
    throw new IndicatorError("RSI requires at least 2 periods", "RSI");
  }

  if (candles.length < 1) {
    throw new IndicatorError("No candles provided for RSI", "RSI");
  }

  const rsiValues = RSI.calculate({
    period: params.periods,
    values: prices,
  });
  const emptyRsiValues = new Array<number>(
    candles.length - rsiValues.length,
  ).fill(NaN);

  return [...emptyRsiValues, ...rsiValues];
}
