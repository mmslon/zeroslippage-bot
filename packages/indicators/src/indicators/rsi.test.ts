import { describe, it, expect } from "vitest";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import type { ICandlestick } from "@opentrader/types";

import { rsi } from "./rsi.js";

const loadCandles = (filename: string): ICandlestick[] => {
  return JSON.parse(readFileSync(join(__dirname, `./__mocks__/${filename}`), "utf-8"));
};

describe("rsi", () => {
  const candles: ICandlestick[] = loadCandles("ETH_USDT-1d-candles.json");
  const candles1h: ICandlestick[] = loadCandles("ETH_USDT-1h-candles.json");

  it("should calculate RSI", async () => {
    const result = await rsi({ periods: 14 }, candles);

    expect(result).toMatchSnapshot();
  });

  it("should throw an error if there a less than 2 periods", async () => {
    await expect(rsi({ periods: 1 }, candles)).rejects.toThrow("RSI requires at least 2 periods");
  });

  it("should throw an error if no candles provided", async () => {
    await expect(rsi({ periods: 14 }, [])).rejects.toThrow("No candles provided for RSI");
  });

  it("the length of RSI values must be equal to the number of candles", async () => {
    const rsiValues = await rsi({ periods: 14 }, candles);
    expect(rsiValues.length).toBe(candles.length);

    const rsiValues2 = await rsi({ periods: 14 }, candles.slice(0, 1));
    expect(rsiValues2.length).toBe(1);
  });

  // it("should start calculating RSI values from the last candle", async () => {
  //   const rsiValues = await rsi({ periods: 7 }, candles.slice(0, 7));
  //   const lastRsiValue = rsiValues[rsiValues.length - 1];
  //
  //   expect(lastRsiValue).not.toBeNaN();
  // });

  it("should calculate RSI for 1h candles", async () => {
    const result = await rsi({ periods: 14 }, candles1h);

    expect(result).toMatchSnapshot();
  });
});
