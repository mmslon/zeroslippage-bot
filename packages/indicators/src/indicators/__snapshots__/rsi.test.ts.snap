// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rsi > should calculate RSI 1`] = `
[
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  66.66,
  57.27,
  57.18,
  57.31,
  60.79,
  61.47,
  52.68,
  61.44,
  61.32,
  62.48,
  62.63,
  60.32,
  56.58,
  58.23,
  56.04,
  58.06,
  61.01,
  66.55,
  68.31,
  71.23,
  73.82,
  66.1,
  72.41,
  72.54,
  70.32,
  70.94,
  56.96,
  54.99,
  59.01,
  62.52,
  53.99,
  54.56,
  51.8,
  53.59,
  49.95,
  52.01,
  55.05,
  61.14,
  59.46,
  55.1,
  55.66,
  51.7,
  62.26,
  59.06,
  55.02,
  54.34,
  53.45,
  58.74,
  59,
  46.31,
  50.81,
  50.91,
  48.65,
  47.07,
  55.81,
  56.76,
  69.37,
  70.66,
  62.61,
  65.1,
  57.29,
  59.24,
  62.8,
  58.63,
  54.59,
  55.75,
  54.35,
  53.24,
  44.09,
  40.35,
  39.97,
  39.09,
  43.09,
  43.13,
  42.38,
  47.63,
  49.66,
  45.18,
  46.97,
  47.41,
  46.37,
  45.76,
  47.07,
  54.04,
  58.4,
  57.77,
  62.92,
  63.89,
  64.39,
  73.3,
  70.83,
  76.62,
  78.27,
  75.76,
  73.83,
  77.54,
  79.6,
  81.63,
  76.13,
  76.28,
  70.6,
  73.62,
  77.83,
  79.65,
  81.39,
  84.43,
  80.14,
  82.24,
  81.08,
  82.59,
  85.26,
  78.4,
  83.53,
  84.27,
  84.49,
  84.84,
  82.42,
  85.53,
  78.7,
  79.23,
  70.06,
  61.37,
  50.72,
  55.33,
  50.19,
  38.78,
  50.68,
  50.01,
  45.58,
  45.38,
  49.52,
  53.62,
  53.52,
  50.54,
  52.49,
  50.6,
  50.44,
  55.5,
  49.98,
  42.64,
  43.91,
  44.6,
  44.24,
  45.8,
  50.26,
  58.86,
  51.35,
  52.71,
  51.05,
  42.19,
  36.26,
  41.9,
  40.54,
  40.05,
  37.43,
  40.78,
  40.54,
  44.82,
  44.53,
  46.89,
  47.73,
  44.53,
  45.31,
  44.25,
  50.54,
  50.91,
  48.55,
  39.88,
  38.35,
  39.21,
  45.78,
  46.55,
  47.63,
  43.99,
  41.38,
  39.92,
  44.01,
  38.31,
  38.47,
  39.7,
  41.26,
  37.85,
  47.99,
  43.53,
  51.66,
]
`;

exports[`rsi > should calculate RSI for 1h candles 1`] = `
[
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  50.24,
  51.04,
  51.82,
  45.91,
  46.15,
  44.02,
  42.33,
  40.89,
  49.39,
  43.42,
]
`;
